system/include/emscripten/emscripten.h

```diff
@@ -255,6 +255,43 @@ typedef void (*em_scan_func)(void*, void*);
 void emscripten_scan_registers(em_scan_func func);
 void emscripten_scan_stack(em_scan_func func);
 
+// Fibers API
+// Requires upstream backend and Asyncify
+
+#define EM_FIBER_ASYNCIFY_STACK_SIZE 4096
+
+typedef struct asyncify_fiber_s {
+    void *stack_base;
+    void *stack_limit;
+    void *stack_ptr;
+    em_arg_callback_func entry;
+    void *user_data;
+} asyncify_fiber_t;
+
+typedef struct asyncify_data_s {
+    void *stack_ptr;
+    void *stack_limit;
+    int rewind_id;
+} asyncify_data_t;
+
+typedef struct emscripten_fiber_s {
+    asyncify_fiber_t asyncify_fiber;
+    asyncify_data_t asyncify_data;
+    char asyncify_stack[EM_FIBER_ASYNCIFY_STACK_SIZE];
+} emscripten_fiber_t;
+
+void emscripten_fiber_init(emscripten_fiber_t *fiber, unsigned int fiber_struct_size, em_arg_callback_func entry_func, void *entry_func_arg, void *stack, unsigned int stack_size);
```

Comment:
Use size_t for sizes?

Reply:
It seems a little strange to pass both the stack size and the struct size.  Can't one be derived from the other?  Seems like just passing stack size would be the cleanest.

Reply:
> Use size_t for sizes?

Yeah, I probably should. For some reason I wasn't sure whether `size_t` is 32-bit or not... Some parts of the API seem to use plain `int` for sizes (old coroutine API included), so that confused me a bit.

> It seems a little strange to pass both the stack size and the struct size. Can't one be derived from the other? Seems like just passing stack size would be the cleanest.

There are two separate stacks involved here: the C stack, and the Asyncify call stack. The stack size parameter is for the C stack, which can be separate from the fiber structure (and in case of the "main fiber", it must be separate). The Asyncify call stack, however, can always be embedded, so it is. The fiber struct can be over- or under-allocated to control the asyncify stack size, though the default should be reasonable for most uses. I know it's a little weird API, but that's the most efficient way I could think of.